cloudprovider/cloud_provider.go

```diff
@@ -0,0 +1,118 @@
+package cloudprovider
+
+import (
+	"encoding/json"
+	"os"
+	"path"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+
+	"github.com/rancher/agent/core/hostInfo"
+	"github.com/rancher/agent/utilities/config"
+)
+
+const (
+	CloudProviderLabel    = "io.rancher.host.provider"
+	RegionLabel           = "io.rancher.host.region"
+	AvailabilityZoneLabel = "io.rancher.host.zone"
+	infoFile              = "info.json"
+	tempFile              = "temp.json"
+)
+
+var (
+	InfoPath = path.Join(config.StateDir(), infoFile)
+	TempPath = path.Join(config.StateDir(), tempFile)
+)
+
+var (
+	providers = make(map[string]Provider)
+)
+
+type Provider interface {
+	Init() error
+	GetHostInfo() (hostInfo.Info, error)
+	ExpireTime() time.Duration
+	Interval() time.Duration
+}
+
+func AddCloudProvider(name string, provider Provider) {
+	if _, exists := providers[name]; exists {
+		logrus.Fatalf("Provider '%s' tried to register twice", name)
+	}
+	logrus.Infof("Provider '%s' adding", name)
+	providers[name] = provider
+}
+
+func GetCloudProviderInfo() {
+	for name, provider := range providers {
+		if err := provider.Init(); err != nil {
+			logrus.Fatalf("Provider '%s' initial failed, err: '%s'", name, err)
+			continue
+		}
+
+		go func(p Provider) {
+			i := 0
+			t0 := time.Now()
+			var err error
```

Comment:
Don't need declaration here. You have to assign these two variables below anyway.